content/download-self.js

```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+// preventing reregistration if reinjected by tabs.executeScript for whatever reason, just in case
+if (typeof self.oldCode !== 'string') {
+  chrome.runtime.onConnect.addListener(port => {
+    if (port.name !== 'downloadSelf') return;
+    const read = r => r.status === 200 ? r.text() : Promise.reject(r.status);
+    const postBack = msg => {
+      port.postMessage(msg);
+      self.oldCode = msg.code;
+    };
+    port.onMessage.addListener(({id, timer}) => {
+      fetch(location.href, {mode: 'same-origin'})
+        .then(read)
+        .then(code => ({id, code: timer && code === self.oldCode ? null : code}))
+        .catch(error => ({id, error}))
```

Comment:
Can we pass error objects through message port? If not then we have to convert the error to a simple object and convert it back (or throw) on receive.

Reply:
No we can't and I we shouldn't as I explained above.

Reply:
This is not related to the previous problem. JavaScript creates an error object when something went wrong including `fetch()`, `.text()`.

Reply:
Like I said, you need to understand why something is used, not just copy it because something else is doing it. The low-level things and libraries use `Error` object also because they must provide all that extra info inside the object like the call stack and a universal interface to testing/examining the error. We don't need it here.